waspc/data/Generator/templates/sdk/wasp/server/operations/wrappers.ts

```diff
@@ -0,0 +1,254 @@
+ * authenticated operation definition types are a subtype of this type).
+ *
+ */
+type GenericAuthenticatedOperationDefinition = AuthenticatedOperationDefinition<
+  // NOTE(filip): Not quite sure I understand what's going on with Variance here.
+  _Entity[],
+  never,
+  Payload
+>
+
+{=/ isAuthEnabled =}
+/**
+ * Constructs the type for an unauthenticated operation's server-side API.
+ *
+ * @template Input The type of the payload the operation expects (must be `void` if the operation doesn't expect a payload).
+ * @template Output The type of the operation's return value.
+ */
+type UnauthenticatedOperation<Input, Output> = Operation<Input, Output, false>
+
+/**
+ * The principal type for an unauthenticated operation's definition (i.e., all
+ * unauthenticated operation definition types are a subtype of this type).
+ *
+ */
+type GenericUnauthenticatedOperationDefinition = UnauthenticatedOperationDefinition<
+  // NOTE(filip): Not quite sure I understand what's going on with Variance here.
+  _Entity[],
+  never,
+  Payload
+>
+
+/**
+ * Queries the entity map from the type of the operation's definition.
+ *
+ * @template OperationDefinition The type of the operation's definition.
+ */
+type EntityMapFor<OperationDefinition extends GenericAuthenticatedOperationDefinition> = 
+  Parameters<OperationDefinition>[1]["entities"]
+
+/**
+ * Constructs the type for an operation's server-side API.
+ *
+ * @template Input The type of the payload the operation expects (must be
+ * `void` if the operation doesn't expect a payload).
+ * @template Output The type of the operation's return value.
+ * @template IsAuthenticated true if the operation is authenticated, false otherwise
+ */
+type Operation<Input, Output, IsAuthenticated extends boolean> = [
+  Input
+] extends [never]
+  ? OperationWithPayload<unknown, Output, IsAuthenticated>
+  : [Input] extends [void]
+  ? OperationWithoutPayload<Output, IsAuthenticated>
+  : OperationWithPayload<Input, Output, IsAuthenticated>
+
+/**
+ * Constructs the type of the server-side API for an operation that
+ * expects a payload.
+ *
+ * @template Input The type of the payload the operation expects.
+ * @template Output The type of the operation's return value.
+ * @template IsAuthenticated true if the operation is authenticated, false otherwise
+ */
+type OperationWithPayload<
+  Input,
+  Output,
+  IsAuthenticated extends boolean
+> = IsAuthenticated extends true
+  ? (args: Input, context: { user: AuthUser }) => Promise<Output>
+  : (args: Input) => Promise<Output>
+
+/**
+ * Constructs the type of the server-side API for an operation that
+ * doesn't expect a payload.
+ *
+ * @template Output The type of the operation's return value.
+ * @template IsAuthenticated true if the operation is authenticated, false otherwise
+ */
+type OperationWithoutPayload<
+  Output,
+  IsAuthenticated extends boolean
+> = IsAuthenticated extends true
```

Comment:
I guess this one of the cases where the boolean type param is necessary 😄  https://github.com/wasp-lang/wasp/pull/1993#discussion_r1632964113

Reply:
Haha, I knew you'll notice this!

Yeah, in this case since:
- It's very local (only one generic calls the boolean)
- The alternative option is writing a Cartesian product with 4 simple generics with duplicate definitions of very subtle behaviors (e.g., optionality of arguments) whose logic is arguably simpler to follow with the boolean.

I've decided to use the boolean.

I might be biased for working inside this file for too long so, if you think the boolean is confusing, I'll gladly remove it.

Reply:
Sounds good, thanks for the detailed write-up 👍